### DISCUSSION TITLE
CCACHE_BASEDIR and out of tree build?

### DISCUSSION THREAD

**Discussion by infn-ke** [ccache/ccache#1344] [Questions & answers]
It seems to me ccache does not support an out of tree build since CCACHE_BASEDIR can only eliminate either source dir or build dir. Is it a requirement that build directory is a sub-directory of the source directory? Or how to build in different directories? Normally an out of tree build looks like this (and folder names source and build are user defined)

```text
/project
    /source
    /build
```

**Comment by jrosdahl** [comment]
It is not a requirement that the build directory is a subdirectory of the source directory, but you are correct that you need a stable name for `source` since otherwise it won't be possible to use the same compilation commands in `build` and another user's project setup where the source directory is named differently. This is regardless of if you use `CCACHE_BASEDIR` to make absolute paths relative or if you have a better build system that uses relative paths from start. You could introduce a naming convention that is shared between the users or perhaps make a `source` symlink (either in `build` or one level up) that points to the real source tree.

Here are some examples of how `CCACHE_BASEDIR` works: https://ccache.dev/manual/4.8.3.html#config_base_dir

**Reply by infn-ke** [reply]
Thanks, for the reply. But would it not be a better solution if ccache instead exposed two variables, one for source dir and one for build dir? That would decouple ccache from how build is setup. Relying on naming conventions enforcing a particular source/build tree layout is fragile.

**Reply by jrosdahl** [reply]
It sounds like you haven't fully grasped what `CCACHE_BASEDIR` does. `CCACHE_BASEDIR` specifies a prefix. If an absolute path matches the prefix it is rewritten as relative to the current working directory, not to `CCACHE_BASEDIR`. Therefore, `CCACHE_BASEDIR` cannot make the name of the source directory just go away.

In your example you can set `CCACHE_BASEDIR` to `/project` and then ccache will rewrite absolute paths to relative if they refer to any path below `/project`, e.g. `/project/source/example.c` or `/project/build/example.o`. If ccache had two different variables and you set one to `/project/source` and one to `/project/build`, the end result would be exactly the same. The thing is that if the current working directory is `/project/build` and you invoke the compiler with `/project/source/example.c` then the path will have to be rewritten to `../source/example.c`, right? So `source` will be part of the path and there is nothing ccache can do about that.

That said, one could imagine setups where it's not optimal to set `CCACHE_BASEDIR` to a single path, e.g. if the layout is like 

```
/aaa/source/example1.c
/bbb/source/example2.c
/ccc/build
/usr/include
```

Then you'd perhaps want to make relative paths of absolute paths below `/aaa` and `/bbb` but not to stuff in `/usr`, so `CCACHE_BASEDIR=/` wouldn't be a good setting and it would be better if you could set multiple base dir paths. That would be quite easy to implement and is tracked by #597. See also #785.
